File: kotlin/0130-surrounded-regions.kt
package kotlin

// Runtime: 356 ms, faster than 72.97% of Kotlin online submissions for Surrounded Regions.
// Memory Usage: 48.5 MB, less than 75.68% of Kotlin online submissions for Surrounded Regions.
// https://leetcode.com/submissions/detail/776042031/
class Solution {
    fun solve(board: Array<CharArray>): Unit {
        val n = board.size
        val m = board[0].size

        for (y in 0 until n) {
            for (x in 0 until m) {
                if (x==0||x==m-1||y==0||y==n-1){
                    if (board[y][x] == 'O') {
                        markI(board, y, x)
                    }
                }
            }
        }

        for (y in 0 until n) {
            for (x in 0 until m) {
                board[y][x] = when(board[y][x]){
                    'O' -> 'X'
                    'I' -> 'O'
                    else -> board[y][x] // could've just 'continue'd only if leetcode uses newer version of Kotlin
                }
            }
        }
    }
    fun inBound(board: Array<CharArray>, r: Int, c: Int) : Boolean {
        val n = board.size
        val m = board[0].size
        return !(r < 0 || r >= n || c < 0 || c >= m)
    }
    fun markI(board: Array<CharArray>, y: Int, x: Int) {
        if (!inBound(board, y, x)) return
        if (board[y][x] != 'O') return
        board[y][x] = 'I'
        markI(board, y+1, x)
        markI(board, y-1, x)
        markI(board, y, x+1)
        markI(board, y, x-1)
    }
}

File: csharp/0130-surrounded-regions.cs
public class Solution {
    public void Solve(char[][] board) {
        var n = board.Length;

        if (n == 0) return;
        var m = board[0].Length;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if ((i == 0 || j == 0 || i == n - 1 || j == m - 1) && board[i][j] == 'O') {
                    CaptureDfs(board, i, j);
                }
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] == 'T') {
                    board[i][j] = 'O';
                }
            }
        }
    }

    private void CaptureDfs(char[][] board, int x, int y) {
        var n = board.Length;
        var m = board[0].Length;

        if (x >= n || x < 0 || y >= m || y < 0) {
            return;
        }

        if (board[x][y] == 'T' || board[x][y] == 'X') return;

        board[x][y] = 'T';
        CaptureDfs(board, x+1, y); 
        CaptureDfs(board, x-1, y); 
        CaptureDfs(board, x, y+1); 
        CaptureDfs(board, x, y-1); 

    }
}
